mono/metadata/gc.c

```diff
@@ -681,6 +682,31 @@ ves_icall_System_GCHandle_GetTarget (guint32 handle)
 	return mono_gchandle_get_target (handle);
 }
 
+static gboolean
+is_object_pinnable (MonoObject *obj)
```

Comment:
On the MONO_TYPE_CLASS check, that's not how to verify if a type is a ref type. It ignores object, generic instances and arrays. You should use MONO_TYPE_IS_REFERENCE.

The array check means you can pin a boxed char, but not an array of chars. I don't believe that's the expected behavior.

As I mentioned, we need positive tests for the types that expected to be pinnable.